### PR TITLE
feat(event_handler): add support to VPC Lattice payload v2

### LINKED ISSUE #3160: Feature request: add support for VPC Lattice v2 in Event Handler, Data Class and parser
### Use case

On Sep 7, 2023, AWS introduced support for VPC Lattice v2 payload and we need to add support in data class/parser/event handler.

Reference: https://docs.aws.amazon.com/vpc-lattice/latest/ug/lambda-functions.html

### Solution/User Experience

@event_parser(model=VPCLatticeV2, envelope=envelopes.VPCLatticeV2)
def handler(event: VPCLatticeV2, context: LambdaContext):
    print(event.body)

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by stephenbawks**
**Issue number:** #3160 

### Changes

As of September 8 there is a new VPC Lattice Event Type for Lambda. 

When you create or update a target group of type LAMBDA, you can specify the version of the event structure that your Lambda function receives. The possible versions are `V1` and `V2`.
https://docs.aws.amazon.com/vpc-lattice/latest/ug/lambda-functions.html

#### Using VPCLatticeV2Resolver Event Handler

```python
import requests
from requests import Response

from aws_lambda_powertools import Logger, Tracer
from aws_lambda_powertools.event_handler import VPCLatticeV2Resolver
from aws_lambda_powertools.logging import correlation_paths
from aws_lambda_powertools.utilities.typing import LambdaContext

tracer = Tracer()
logger = Logger()
app = VPCLatticeV2Resolver()


@app.get("/todos")
@tracer.capture_method
def get_todos():
    todos: Response = requests.get("https://jsonplaceholder.typicode.com/todos")
    todos.raise_for_status()

    # for brevity, we'll limit to the first 10 only
    return {"todos": todos.json()[:10]}


# You can continue to use other utilities just as before
@logger.inject_lambda_context(correlation_id_path=correlation_paths.APPLICATION_LOAD_BALANCER)
@tracer.capture_lambda_handler
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

#### Using VPCLatticeEventV2 Data Class

```python
from aws_lambda_powertools import Logger
from aws_lambda_powertools.utilities.data_classes import VPCLatticeEventV2, event_source
from aws_lambda_powertools.utilities.typing import LambdaContext

logger = Logger()


@event_source(data_class=VPCLatticeEventV2)
def lambda_handler(event: VPCLatticeEventV2, context: LambdaContext):
    logger.info(event.body)

    response = {
        "isBase64Encoded": False,
        "statusCode": 200,
        "statusDescription": "200 OK",
        "headers": {"Content-Type": "application/text"},
        "body": "VPC Lattice V2 Event ✨🎉✨",
    }

    return response
```

#### Using VPCLatticeEventV2 Parser

```python
from aws_lambda_powertools.utilities.parser import event_parser
from aws_lambda_powertools.utilities.parser.models import VpcLatticeV2Model
from aws_lambda_powertools.utilities.typing import LambdaContext

@event_parser(model=VpcLatticeV2Model)
def handler(event: dict, context: LambdaContext):
    body = event.body

    assert body is not None

```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [ ] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Hi @stephenbawks! I believe there is some testing and documentation left to be written about this new event source data class. Can you check it please?

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Missing tests before reviewing this.

**Comment by leandrodamascena** [comment]
I see the CI failed because Ruff is complaining. It's easy to fix :D

Thanks for updating this event handler, @stephenbawks! You always tune in to new AWS features.

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py [hunk 1]
```diff
@@ -4,15 +4,15 @@
...
+from aws_lambda_powertools.utilities.data_classes.common import BaseProxyEvent, DictWrapper
 from aws_lambda_powertools.utilities.data_classes.shared_functions import (
     base64_decode,
     get_header_value,
     get_query_string_value,
 )
 
 
-class VPCLatticeEvent(BaseProxyEvent):
+class VPCLatticeEventBase(BaseProxyEvent):
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:15]: Many fields, decode and others are shared by both events, that's why I created this base class.

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py [hunk 2]
```diff
@@ -140,3 +122,75 @@ def get_header_value(
...
+
+    @property
+    def path(self) -> str:
+        """The VPC Lattice v2 Event path"""
+        return self["path"]
+
+    @property
+    def request_context(self) -> vpcLatticeEventV2RequestContext:
+        """he VPC Lattice v2 Event request context."""
+        return vpcLatticeEventV2RequestContext(self["requestContext"])
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:248]: Was missing this field.

File: aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    time_epoch: datetime = Field(alias="timeEpoch")
+
+
+class VpcLatticeV2Model(BaseModel):
+    version: str
+    path: str
+    method: str
+    headers: Dict[str, str]
+    query_string_parameters: Optional[Dict[str, str]] = None
+    body: Optional[Union[str, Type[BaseModel]]] = None
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py:40]: When using GET this is optional.

File: aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    is_base64_encoded: Optional[bool] = Field(None, alias="isBase64Encoded")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py:41]: When using GET or plain text payload (json, text and others) this is optional.

File: tests/unit/data_classes/test_vpc_lattice_eventv2.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    assert parsed_event.headers == raw_event["headers"]
+    assert parsed_event.query_string_parameters == raw_event["query_string_parameters"]
+    assert parsed_event.body == raw_event["body"]
+    assert parsed_event.is_base64_encoded == raw_event["isBase64Encoded"]
+    assert parsed_event.request_context.region == raw_event["requestContext"]["region"]
+    assert parsed_event.request_context.service_network_arn == raw_event["requestContext"]["serviceNetworkArn"]
+    assert parsed_event.request_context.service_arn == raw_event["requestContext"]["serviceArn"]
+    assert parsed_event.request_context.target_group_arn == raw_event["requestContext"]["targetGroupArn"]
+    assert parsed_event.request_context.source_vpc_arn == raw_event["requestContext"]["identity"]["sourceVpcArn"]
+    assert parsed_event.request_context.time_epoch == raw_event["requestContext"]["timeEpoch"]
```
Comment (leandrodamascena) [inline tests/unit/data_classes/test_vpc_lattice_eventv2.py]: Added more fields.

**Review by leandrodamascena — COMMENTED** [review]
Hey @stephenbawks! I left some comments where I changed things.

Still missing small changes in parser and we are good to merge.

File: tests/unit/data_classes/test_vpc_lattice_eventv2.py (hunk 5 above)
Comment (stephenbawks) [reply tests/unit/data_classes/test_vpc_lattice_eventv2.py]: Ahh nice, I was trying to see if there was a definitive list of fields that would come back.

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py [hunk 6]
```diff
@@ -140,3 +122,127 @@ def get_header_value(
...
+        return self["is_base64_encoded"]
+
+    # VPCLattice event has no path field
+    # Added here for consistency with the BaseProxyEvent class
+    @property
+    def path(self) -> str:
+        return self["raw_path"]
+
+
+class vpcLatticeEventV2Identity(DictWrapper):
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:145]: I added missing fields.

Look at this documentation - https://docs.aws.amazon.com/lambda/latest/dg/services-vpc-lattice.html#vpc-lattice-receiving-events

```json
"requestContext": {
        "serviceNetworkArn": "arn:aws:vpc-lattice:region:123456789012:servicenetwork/sn-0bf3f2882e9cc805a",
        "serviceArn": "arn:aws:vpc-lattice:region:123456789012:service/svc-0a40eebed65f8d69c",
        "targetGroupArn": "arn:aws:vpc-lattice:region:123456789012:targetgroup/tg-6d0ecf831eec9f09",
        "identity": {
            "sourceVpcArn": "arn:aws:ec2:region:123456789012:vpc/vpc-0b8276c84697e7339",
            "type" : "AWS_IAM",
            "principal": "arn:aws:sts::123456789012:assumed-role/example-role/057d00f8b51257ba3c853a0f248943cf",
            "sessionName": "057d00f8b51257ba3c853a0f248943cf",
            "x509SanDns": "example.com"
        }
        "region": "region",
        "timeEpoch": "1690497599177430"
    }
```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
We need to review the event handler with this new payload. Do you want to add tests or do I?

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py (hunk 6 above)
Comment (stephenbawks) [reply aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:145]: Looks like the documentation is missing a comma on the line right before the region.  

I will get a new test created.

**Comment by leandrodamascena** [comment]
Hey @heitorlessa! Can you review this PR, please?

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py (hunk 1 above)
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:15]: Use Base as a prefix to ease search and standards.

```suggestion
class BaseVPCLatticeEvent(BaseProxyEvent):
```

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py [hunk 7]
```diff
@@ -140,3 +117,137 @@ def get_header_value(
...
+
+    @property
+    def region(self) -> str:
+        """The VPC Lattice v2 Event requestContext serviceNetworkArn"""
+        return self["region"]
+
+    @property
+    def time_epoch(self) -> float:
+        """The VPC Lattice v2 Event requestContext timeEpoch"""
+        return self["timeEpoch"]
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:226]: Are we converting it? IIRC you showed me the sample event and they send it as a str 🤦‍♂️

File: aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+class VpcLatticeV2RequestContext(BaseModel):
+    service_network_arn: str = Field(alias="serviceNetworkArn")
+    service_arn: str = Field(alias="serviceArn")
+    target_group_arn: str = Field(alias="targetGroupArn")
+    identity: VpcLatticeV2RequestContextIdentity
+    region: str
+    time_epoch: float = Field(alias="timeEpoch")
+    time_epoch_as_datetime: datetime = Field(alias="timeEpoch")
+
+    @validator("time_epoch_as_datetime", pre=True, allow_reuse=True)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py:29]: Add a comment as to why this is needed so we don't forget.

I think there's something wrong here as the validator name doesn't match the logic - leftover perhaps?

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py:29]: Weren't you gonna add a new property that returns datetime?

File: docs/utilities/data_classes.md [hunk 9]
```diff
@@ -1180,7 +1181,25 @@ AWS Secrets Manager rotation uses an AWS Lambda function to update the secret. [
             do_something_with(record.body)
     ```
 
-### VPC Lattice
+### VPC Lattice V2
```
Comment (heitorlessa) [inline docs/utilities/data_classes.md:1184]: Suggestion - keep without the v2 to not break deep linking any customer might have had.

The default will be v2 then, and V1 gets a new anchor

**Review by heitorlessa — CHANGES_REQUESTED** [review]

File: aws_lambda_powertools/utilities/data_classes/vpc_lattice.py (hunk 7 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/data_classes/vpc_lattice.py:226]: You're right, I made a mistake.

File: aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py (hunk 8 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parser/models/vpc_latticev2.py:29]: I changed the name of this validator and added a comment.

**Comment by leandrodamascena** [comment]
Hey @heitorlessa! I updated the PR body and addressed your feedback. Can you review it again, please?

**Review by heitorlessa — APPROVED** [review]

**Review by leandrodamascena — APPROVED** [review]

**Pull request merged**